subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedFeatureUsage.java

```diff
@@ -48,41 +48,45 @@ public String getMessage() {
         return stack;
     }
 
-    /**
-     * Creates a copy of this usage with the stack trace populated. Implementation is a bit limited in that it assumes that
-     * this method is called from the same thread that triggered the usage.
-     */
-    public DeprecatedFeatureUsage withStackTrace() {
-        if (!stack.isEmpty()) {
-            return this;
-        }
-
+    private static List<StackTraceElement> createStackTrace(Class<?> calledFrom) {
         StackTraceElement[] originalStack = new Exception().getStackTrace();
-        int caller = 0;
-        while (caller < originalStack.length && !originalStack[caller].getClassName().startsWith(calledFrom.getName())) {
-            caller++;
-        }
-        while (caller < originalStack.length && originalStack[caller].getClassName().startsWith(calledFrom.getName())) {
-            caller++;
+        final String calledFromName = calledFrom.getName();
+        boolean calledFromFound = false;
+        int caller;
+        for (caller = 0; caller < originalStack.length; caller++) {
+            StackTraceElement current = originalStack[caller];
+            if (!calledFromFound) {
+                if (current.getClassName().startsWith(calledFromName)) {
+                    calledFromFound = true;
+                }
+            } else {
+                if (!current.getClassName().startsWith(calledFromName)) {
+                    break;
+                }
+            }
         }
+
         caller = skipSystemStackElements(originalStack, caller);
-        caller++;
-        caller = skipSystemStackElements(originalStack, caller);
-        List<StackTraceElement> stack = new ArrayList<StackTraceElement>();
+
+        List<StackTraceElement> result = new ArrayList<StackTraceElement>();
         for (; caller < originalStack.length; caller++) {
-            stack.add(originalStack[caller]);
+            result.add(originalStack[caller]);
         }
-        return new DeprecatedFeatureUsage(this, stack);
+
+        return result;
     }
 
-    private int skipSystemStackElements(StackTraceElement[] stackTrace, int caller) {
-        while (caller < stackTrace.length && stackTrace[caller].getClassName().startsWith("org.codehaus.groovy.")
-                || stackTrace[caller].getClassName().startsWith("org.gradle.internal.metaobject.")
-                || stackTrace[caller].getClassName().startsWith("groovy.")
-                || stackTrace[caller].getClassName().startsWith("java.")
-                || stackTrace[caller].getClassName().startsWith("sun.")
-                || stackTrace[caller].getClassName().startsWith("jdk.internal.")) {
-            caller++;
+    private static int skipSystemStackElements(StackTraceElement[] stackTrace, int caller) {
+        for (; caller < stackTrace.length; caller++) {
```

Comment:
I mainly dislike repeated identical code like `stackTrace[caller].getClassName()` and would always introduce a local variable in such a case.

So I needed the scope...

And since the body of the loop was incrementing `caller` anyway... :)